File: backend (hms)/doctor-management-service/src/main/java/com/pro/doctormanagementservice/service/DoctorService.java
package com.pro.doctormanagementservice.service;

import com.pro.doctormanagementservice.config.JwtUtils;
import com.pro.doctormanagementservice.dto.DoctorDTO;
import com.pro.doctormanagementservice.dto.CreateDoctorDTO;
import com.pro.doctormanagementservice.dto.UpdateDoctorDTO;
import com.pro.doctormanagementservice.exception.ResourceNotFoundException;
import com.pro.doctormanagementservice.mapper.DoctorMapper;
import com.pro.doctormanagementservice.model.Doctor;
import com.pro.doctormanagementservice.repository.DoctorRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.security.core.Authentication;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Transactional
public class DoctorService {

    private final DoctorRepository doctorRepository;
    private final DoctorMapper doctorMapper;
    private final JwtUtils jwtUtils;

    public List<DoctorDTO> getAllDoctors() {
        return doctorRepository.findAll().stream()
                .map(doctorMapper::toDto)
                .collect(Collectors.toList());
    }

    public DoctorDTO getDoctorById(Long id) {
        return doctorRepository.findById(id)
                .map(doctorMapper::toDto)
                .orElseThrow(() -> new ResourceNotFoundException("Doctor not found with id: " + id));
    }

    public DoctorDTO getDoctorByUserId(Long userId) {
        // Since doctor ID and user ID are the same, delegate to getDoctorById
        return getDoctorById(userId);
    }

    public DoctorDTO createDoctor(CreateDoctorDTO createDoctorDTO) {
        // Manually create the Doctor entity to avoid MapStruct ID mapping issues
        Doctor doctor = Doctor.builder()
                .firstName(createDoctorDTO.getFirstName())
                .lastName(createDoctorDTO.getLastName())
                .email(createDoctorDTO.getEmail())
                .mobile(createDoctorDTO.getMobile())
                .licenseNumber(createDoctorDTO.getLicenseNumber())
                .specialization(createDoctorDTO.getSpecialization())
                .experienceYears(createDoctorDTO.getExperienceYears())
                .qualification(createDoctorDTO.getQualification())
                .dateOfBirth(createDoctorDTO.getDateOfBirth())
                .gender(createDoctorDTO.getGender())
                .hireDate(createDoctorDTO.getHireDate())
                .status(createDoctorDTO.getStatus())
                .consultationFee(createDoctorDTO.getConsultationFee())
                .bio(createDoctorDTO.getBio())
                .build();
        
        // Let JPA auto-generate the ID and timestamps
        doctor = doctorRepository.save(doctor);
        return doctorMapper.toDto(doctor);
    }

    public DoctorDTO updateDoctor(Long id, UpdateDoctorDTO updateDoctorDTO) {
        Doctor doctor = doctorRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Doctor not found with id: " + id));

        doctorMapper.updateEntityFromDto(updateDoctorDTO, doctor);
        doctor = doctorRepository.save(doctor);
        return doctorMapper.toDto(doctor);
    }

    public void deleteDoctor(Long id) {
        if (!doctorRepository.existsById(id)) {
            throw new ResourceNotFoundException("Doctor not found with id: " + id);
        }
        doctorRepository.deleteById(id);
    }

    /**
     * Check if the authenticated user is the doctor with the given doctor ID
     * Since doctor.id now directly represents the user ID from auth service
     */
    public boolean isDoctorOwnProfile(Long doctorId, Authentication authentication) {
        try {
            Long authenticatedUserId = jwtUtils.getUserIdFromAuthentication(authentication);
            if (authenticatedUserId == null) {
                return false;
            }

            // Since doctor.id is the user ID, we can directly compare
            return Objects.equals(doctorId, authenticatedUserId);
        } catch (Exception e) {
            return false;
        }
    }
} 

File: frontend/client/src/services/doctorService.ts
import { api } from '@/lib/api';
import { Doctor, UpdateDoctorRequest, Specialization, Gender, DoctorStatus } from '@/components/types/doctor';

export interface CreateDoctorRequest {
  firstName: string;
  lastName: string;
  email: string;
  mobile: string;
  licenseNumber: string;
  specialization: Specialization;
  experienceYears: number;
  qualification: string;
  dateOfBirth: string; // ISO date string
  gender: Gender;
  hireDate: string; // ISO date string
  status: DoctorStatus;
  consultationFee: number;
  bio?: string;
}

/**
 * Doctor Service
 * Handles all API calls related to doctor management
 */
export class DoctorService {
  private static readonly BASE_URL = '/doctors';

  /**
   * Get doctor information by user ID
   * @param userId - The user ID to fetch doctor data for
   * @returns Promise<Doctor>
   */
  static async getDoctorByUserId(userId: number): Promise<Doctor> {
    try {
      const response = await api.get(`${this.BASE_URL}/user/${userId}`);
      return response.data;
    } catch (error: any) {
      console.error('Error fetching doctor by user ID:', error);
      throw new Error(
        error.response?.data?.message || 
        'Failed to fetch doctor information'
      );
    }
  }

  /**
   * Get doctor information by doctor ID
   * @param doctorId - The doctor ID to fetch data for
   * @returns Promise<Doctor>
   */
  static async getDoctorById(doctorId: number): Promise<Doctor> {
    try {
      const response = await api.get(`${this.BASE_URL}/${doctorId}`);
      return response.data;
    } catch (error: any) {
      console.error('Error fetching doctor by ID:', error);
      throw new Error(
        error.response?.data?.message || 
        'Failed to fetch doctor information'
      );
    }
  }

  /**
   * Update doctor information
   * @param doctorId - The doctor ID to update
   * @param updateData - The data to update
   * @returns Promise<Doctor>
   */
  static async updateDoctor(doctorId: number, updateData: UpdateDoctorRequest): Promise<Doctor> {
    try {
      const response = await api.put(`${this.BASE_URL}/${doctorId}`, updateData);
      return response.data;
    } catch (error: any) {
      console.error('Error updating doctor:', error);
      throw new Error(
        error.response?.data?.message || 
        'Failed to update doctor information'
      );
    }
  }

  /**
   * Get all doctors (admin only)
   * @returns Promise<Doctor[]>
   */
  static async getAllDoctors(): Promise<Doctor[]> {
    try {
      const response = await api.get(`${DoctorService.BASE_URL}`);
      return response.data;
    } catch (error: any) {
      console.error('Error fetching all doctors:', error);
      throw new Error(
        error.response?.data?.message || 
        'Failed to fetch doctors list'
      );
    }
  }

  /**
   * Create a new doctor (admin only)
   * @param createData - The data for creating a new doctor
   * @returns Promise<Doctor>
   */
  static async createDoctor(createData: CreateDoctorRequest): Promise<Doctor> {
    try {
      const response = await api.post(this.BASE_URL, createData);
      return response.data;
    } catch (error: any) {
      console.error('Error creating doctor:', error);
      throw new Error(
        error.response?.data?.message || 
        'Failed to create doctor'
      );
    }
  }

  /**
   * Delete a doctor (admin only)
   * @param doctorId - The doctor ID to delete
   * @returns Promise<void>
   */
  static async deleteDoctor(doctorId: number): Promise<void> {
    try {
      await api.delete(`${this.BASE_URL}/${doctorId}`);
    } catch (error: any) {
      console.error('Error deleting doctor:', error);
      throw new Error(
        error.response?.data?.message || 
        'Failed to delete doctor'
      );
    }
  }
}

export default DoctorService; 